beluga/include/beluga/sensor/likelihood_3d_field_model.hpp

```diff
@@ -0,0 +1,167 @@
+
+#include <Eigen/Core>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/transform.hpp>
+#include <sophus/se3.hpp>
+#include <sophus/so3.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a likelihood field sensor model for 3D Lidars.
+ */
+
+namespace beluga {
+
+/// Parameters used to construct a Likelihood3DFieldModel instance.
+/**
+ * See Probabilistic Robotics \cite thrun2005probabilistic Chapter 6.4, particularly Table 6.3.
+ */
+struct Likelihood3DFieldModelParam {
+  /// Voxel size in meters.
+  double voxel_size = 0.07;
+  /// Maximum distance to obstacle.
+  /**
+   * When creating a distance map, if the distance to an obstacle is higher than the value specified here,
+   * then this value will be used.
+   */
+  double max_obstacle_distance = 100.0;
+  /// Maximum range of a laser ray.
+  double max_laser_distance = 2.0;
+  /// Weight used to combine the probability of hitting an obstacle.
+  double z_hit = 0.5;
+  /// Weight used to combine the probability of random noise in perception.
+  double z_random = 0.5;
+  /// Standard deviation of a gaussian centered arounds obstacles.
+  /**
+   * Used to calculate the probability of the obstacle being hit.
+   */
+  double sigma_hit = 0.2;
+};
+
+/// Likelihood field sensor model for range finders.
+/**
+ * This model relies on a pre-computed likelihood map of the environment.
+ * It is less computationally intensive than the beluga::BeamSensorModel
+ * because no ray-tracing is required, and it can also provide better
+ * performance in environments with non-smooth occupation maps. See
+ * Probabilistic Robotics \cite thrun2005probabilistic, Chapter 6.4,
+ * for further reference.
+ *
+ * \note This class satisfies \ref SensorModelPage.
+ *
+ * \tparam OpenVDB grid type.
+ */
+template <typename GridT, typename U>
+class Likelihood3DFieldModel {
+ public:
+  /// State type of a particle.
+  using state_type = Sophus::SE3d;
+
+  /// Weight type of the particle.
+  using weight_type = double;
+
+  /// Measurement type given by the interface.
+  using measurement_type = U;
+
+  /// Map representation type.
+  using map_type = GridT;
+
+  /// Parameter type that the constructor uses to configure the likelihood field model.
+  using param_type = Likelihood3DFieldModelParam;
+
+  /// Constructs a Likelihood3DFieldModel instance.
+  /**
+   * \param params Parameters to configure this instance.
+   *  See beluga::Likelihood3DFieldModelParam for details.
+   * \param grid Narrow band Level set grid representing the static map that the sensor model
+   *  uses to compute a likelihood field for lidar hits and compute importance weights
+   *  for particle states.
+   *  Currently only supports OpenVDB Level sets.
+   */
+  explicit Likelihood3DFieldModel(const param_type& params, const map_type& grid)
+      : params_{params},
+        looker_{openvdb::tools::ClosestSurfacePoint<map_type>::create(grid)},
```

Comment:
@pvela2017 shall we have an `update_map()` method that takes care of creating the `ClosestSurfacePoint` instance from a given `grid`?

Reply:
I think no needed yet